### PR TITLE
[web] Resolving all redirected modules to `expo`

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Webpack currently takes ~70 seconds to parse all of the modules, the complex redirects were adding ~3-5 seconds. Cut down on unneeded redirects and optimized the existing ones so the initial webpack time would be reduced. 

# How

* Removed extra redirects
* Added web Platform to expo/web
* Created better `AssetSourceResolver`, and other image utils
* Fixed Linking bug

### CODE REVIEW

File: packages/expo/src/web/Image/AssetSourceResolver.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
+import { Platform } from 'expo-core';
```
Comment (ide) [inline packages/expo/src/web/Image/AssetSourceResolver.ts]: Put .web.ts extensions on all these files so it's clearer that they should not be imported outside of web

File: packages/expo/src/web/Image/AssetSourceResolver.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import invariant from 'invariant';
+import { Dimensions } from 'react-native';
+
+import {
+  getAndroidResourceFolderName,
```
Comment (ide) [inline packages/expo/src/web/Image/AssetSourceResolver.ts]: Remove all this Android stuff -- keep just what's needed for web

File: apps/native-component-list/webpack.config.js [hunk 3]
```diff
@@ -226,29 +221,31 @@ module.exports = {
...
-    ...useWebModules(
-      [
-        'Performance/Systrace',
-        ['HMRLoadingView', 'Utilities/HMRLoadingView'],
-        ['RCTNetworking', 'Network/RCTNetworking'],
-      ],
-      'expo/build/web/'
-    ),
+
+    useWebModule('Platform', 'Utilities/Platform'),
```
Comment (ide) [inline apps/native-component-list/webpack.config.js]: Does this override both react-native and expo-core's Platform?

Comment (EvanBacon) [reply apps/native-component-list/webpack.config.js]: This takes any instance of Platform used internally in react-native and redirects it to use the expo/web version

File: packages/expo/src/web/Image/AssetSourceResolver.ts (hunk 1 above)
Comment (EvanBacon) [reply packages/expo/src/web/Image/AssetSourceResolver.ts]: TS seems to have a problem with this. These files are agnostic shims which aren't ever referenced directly. Ideally we would move this `web/` folder into another library

Comment (ide) [reply packages/expo/src/web/Image/AssetSourceResolver.ts]: My understanding is that this file is "swapped in" with the webpack alias rule -> I don't quite understand what this does to TS, could you explain a bit more?

File: packages/expo/src/web/Image/AssetSourceResolver.ts (hunk 2 above)
Comment (ide) [reply packages/expo/src/web/Image/AssetSourceResolver.ts]: There's maybe very little used on web if we remove all the iOS & Android stuff -> the bundle size can be smaller + we have less surface area to maintain

File: apps/native-component-list/webpack.config.js (hunk 3 above)
Comment (ide) [reply apps/native-component-list/webpack.config.js]: What do you think about adding `select()` to expo-core's Platform and then having web/utilities/Platform export that? That way we have a consistent Platform everywhere.

File: packages/expo/src/web/Image/AssetSourceResolver.ts (hunk 1 above)
Comment (EvanBacon) [reply packages/expo/src/web/Image/AssetSourceResolver.ts]: If there is no default file ex: `foo.ts` then I cannot `import 'foo'`

File: apps/native-component-list/webpack.config.js (hunk 3 above)
Comment (EvanBacon) [reply apps/native-component-list/webpack.config.js]: Platform has a couple more features. In order to really fix this we'd need to PR a web version into RN, or pull out `Platform` into another library that has web support.

Comment (ide) [reply apps/native-component-list/webpack.config.js]: OK, it seems like something worth doing eventually

File: packages/expo/src/web/Image/AssetSourceResolver.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    return !!(this.jsbundleUrl && this.jsbundleUrl.startsWith('file://'));
+  }
+  defaultAsset(): ResolvedAssetSource {
+    if (this.isLoadedFromServer()) {
+      return this.assetServerURL();
+    }
+
+    return this.scaledAssetURLNearBundle();
+  }
+  assetServerURL(): ResolvedAssetSource {
```
Comment (ide) [inline packages/expo/src/web/Image/AssetSourceResolver.ts]: Stuff like this seems unused -> maybe add a console.warn to each method and get an idea of what is called + grep to make a final decision about what needs to stay around.

**Pull request merged**